### ISSUE TITLE
The appliance is offline - But it is not

### ISSUE THREAD

**Issue by DJay-X** [thoukydides/homebridge-homeconnect#19]
Hi...

recently I get a lot of errors in the log. v0.17.0 
Siri confirms that the coffee maker is turned off but it isn't. 
Or Siri points that the device is not responding.
Both is strange.

With the Home Connect App I still can control the machine and it responds instantly.

Any Idea?
Thanks in advance.

```
Feb 23 19:16:48 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:16:48] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:16:48 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:16:48] [HomeConnect] Starting events stream for all appliances
Feb 23 19:17:19 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:17:19] [HomeConnect] [Kaffeevollautomat] SET Off
Feb 23 19:17:19 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:17:19] [HomeConnect] [Kaffeevollautomat] The appliance is offline
Feb 23 19:17:19 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:17:19] [HomeConnect] [Kaffeevollautomat] SET BSH.Common.Setting.PowerState=BSH.Common.EnumType.PowerState.Standby
Feb 23 19:17:19 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:17:19] [HomeConnect] [Kaffeevollautomat] SET Off
Feb 23 19:17:19 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:17:19] [HomeConnect] [Kaffeevollautomat] The appliance is offline
Feb 23 19:17:19 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:17:19] [HomeConnect] [Kaffeevollautomat] SET BSH.Common.Setting.PowerState=BSH.Common.EnumType.PowerState.Standby
Feb 23 19:18:00 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:18:00] [HomeConnect] [Kaffeevollautomat] SET Off
Feb 23 19:18:00 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:18:00] [HomeConnect] [Kaffeevollautomat] The appliance is offline
Feb 23 19:18:00 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:18:00] [HomeConnect] [Kaffeevollautomat] SET BSH.Common.Setting.PowerState=BSH.Common.EnumType.PowerState.Standby
Feb 23 19:18:00 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:18:00] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:18:00 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:18:00] [HomeConnect] Starting events stream for all appliances
Feb 23 19:18:12 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:18:12] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:18:12 raspberrypi-3BPlus homebridge[8966]: [2020-2-23 19:18:12] [HomeConnect] Starting events stream for all appliances
```

```
Feb 23 19:22:28 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:22:28] [HomeConnect] Failed to read list of home appliances: StatusCodeError: Home Connect API error: Timeout on Home Connect subsystem. Please try it again. [SDK.Error.504.GatewayTimeout]
Feb 23 19:22:28 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:22:28] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:22:28 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:22:28] [HomeConnect] Starting events stream for all appliances
Feb 23 19:22:40 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:22:40] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:22:40 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:22:40] [HomeConnect] Starting events stream for all appliances
Feb 23 19:22:52 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:22:52] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:22:52 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:22:52] [HomeConnect] Starting events stream for all appliances
Feb 23 19:23:05 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:23:05] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:23:05 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:23:05] [HomeConnect] Starting events stream for all appliances
Feb 23 19:23:17 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:23:17] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:23:17 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:23:17] [HomeConnect] Starting events stream for all appliances
Feb 23 19:23:29 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:23:29] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:23:29 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:23:29] [HomeConnect] Starting events stream for all appliances
Feb 23 19:23:41 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:23:41] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:23:41 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:23:41] [HomeConnect] Starting events stream for all appliances
Feb 23 19:23:53 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:23:53] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:23:53 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:23:53] [HomeConnect] Starting events stream for all appliances
Feb 23 19:24:05 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:24:05] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:24:05 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:24:05] [HomeConnect] Starting events stream for all appliances
Feb 23 19:24:18 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:24:18] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:24:18 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:24:18] [HomeConnect] Starting events stream for all appliances
Feb 23 19:24:30 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:24:30] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:24:30 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:24:30] [HomeConnect] Starting events stream for all appliances
Feb 23 19:24:42 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:24:42] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:24:42 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:24:42] [HomeConnect] Starting events stream for all appliances
Feb 23 19:24:54 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:24:54] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:24:54 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:24:54] [HomeConnect] Starting events stream for all appliances
Feb 23 19:25:06 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:25:06] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:25:06 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:25:06] [HomeConnect] Starting events stream for all appliances
Feb 23 19:25:18 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:25:18] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:25:18 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:25:18] [HomeConnect] Starting events stream for all appliances
Feb 23 19:25:31 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:25:31] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:25:31 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:25:31] [HomeConnect] Starting events stream for all appliances
Feb 23 19:25:43 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:25:43] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:25:43 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:25:43] [HomeConnect] Starting events stream for all appliances
Feb 23 19:25:55 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:25:55] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:25:55 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:25:55] [HomeConnect] Starting events stream for all appliances
Feb 23 19:26:07 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:26:07] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:26:07 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:26:07] [HomeConnect] Starting events stream for all appliances
Feb 23 19:26:18 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:26:18] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Proxy Error
Feb 23 19:26:18 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:26:18] [HomeConnect] Starting events stream for all appliances
Feb 23 19:26:30 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:26:30] [HomeConnect] Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
Feb 23 19:26:30 raspberrypi-3BPlus homebridge[6844]: [2020-2-23 19:26:30] [HomeConnect] Starting events stream for all appliances
```

**Comment by thoukydides** [comment]
Those are all errors with the Home Connect API. They are not caused by this plugin, and there is nothing that I can do about them.

I have knocked together a quick Home Connect API status monitor which you can find at:
https://www.thouky.co.uk/homeconnect.html

Obviously this is running via my own internet connection and monitoring my own appliances so it won't show problems that only affect some users, and will have some false alarms if there are problems with my network connection. However, in this case you will see that I experienced similar error messages around the same time that you did.

**Comment by DJay-X** [comment]
Oh man.. sorry to bother you. To me it seems that the Home Connect API is very unreliable. 

My log is flooded now with the error every 10 sec. 

```
Starting events stream for all appliances
Terminated events stream for all appliances: Home Connect API error: Gateway Timeout
```

But you Home Connect API status monitor will help to see, that the problem is "global" and not due to anything local or with the plugin.

Again, sorry to bother.

**Comment by thoukydides** [comment]
Yes, the API is quite unreliable. I have suggested to the Home Connect team that there should be an official status page but haven’t received a response yet... which is why I created my own unofficial version.

**Comment by DJay-X** [comment]
@thoukydides Out of context.. but I noticed when I restart homebridge and the coffee maker is off, because of the restart the coffee maker is turning on every time. Is this meant to be?

**Comment by thoukydides** [comment]
> I noticed when I restart homebridge and the coffee maker is off, because of the restart the coffee maker is turning on every time. Is this meant to be?

This is the fix for issue #17. Due to poor design/implementation of the Home Connect API some appliances do not return the correct options for programs unless they are currently selected. Hence, to construct the configuration schema it is necessary to turn on the appliance, select each program in turn, and read their options.

The results should be cached, so this should only happen the first time that the plugin is started after upgrading to v0.17.0, when the `Identify` routine is invoked, or when the cache is invalidated (e.g. by changing the API language). This is how it behaves with my physical appliances and with the appliance simulators.

If you are seeing this behaviour every time that you restart the plugin then please raise a new issue for it and provide a full debug log.
